# Importing libraries

In [1]:
import pandas as pd

# Stage 1: Data mining 

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except FileNotFoundError:
    data = pd.read_csv(r'C:\Users\ASUS\Desktop\Практикум\Data\data.csv')

Let's take a look at the dataset

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


# Stage 2: Data preprocessing 

Removing gaps

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Gaps in "days_employed" will be processed in the next step.

The amount of income is most influenced by the type of employment, so you need to fill in the gaps in this column with the median value for each type from the "income_type" column.

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

Handling anomalous values

In [7]:
data['days_employed'] = data['days_employed'].abs()

For each type of employment, we will display the median value of work experience “days_employed” in days.

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

Two types (безработный (unemployed) and пенсионер (pensioners)) will have abnormally large values. These values are difficult to correct, so leave them as they are. Moreover, we don't need this column for research.

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

There are two anomalous values in the "children" column

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

Let's fill in the gaps in the "days_employed" column with the median values for each type of employment "income_type"

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Changing Data Types

In [14]:
data['total_income'] = data['total_income'].astype(int)

Handling duplicates

In [15]:
data['education'] = data['education'].str.lower()

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

Data categorization

Let's create a column "total_income_category" in the data frame with categories based on ranges:

* 0–30000 - 'E';
* 30001–50000 - 'D';
* 50001–200000 - 'C';
* 200001–1000000 - 'B';
* 1000001 and above - 'A'.

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Let's display a list of unique purposes for taking out a loan from the "purpose" column

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Let's create a function that, based on the data from the "purpose" column, will form a new "purpose_category" column, which will include the following categories:

* 'операции с автомобилем'  (car operations)
* 'операции с недвижимостью'  (real estate transactions)
* 'проведение свадьбы'  (carrying out a wedding)
* 'получение образования'  (getting an education)



In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

# Stage 2: Explore the data and answer questions

### Is there a relationship between the number of children and repayment of the loan on time?

In [23]:

for i in range(6):
    children_debt = round(100*data['debt'].loc[data['children']== i].mean(),2)
    clients_groped = len(data.loc[data['children']==i])
    clients_share = round(100*len(data.loc[data['children']==i])/len(data),2),'%'
    if i == 0:
        print("Out of",clients_groped,clients_share, "clients without children, the percentage of non-returns on time is equal to:" ,children_debt ,"%")
    else:
        print('Out of',clients_groped,clients_share, "clients with a loan arrear, having", i ,"children, do not return the loan on time" ,children_debt ,"%")    


Out of 14091 (66.06, '%') clients without children, the percentage of non-returns on time is equal to: 7.54 %
Out of 4808 (22.54, '%') clients with a loan arrear, having 1 children, do not return the loan on time 9.23 %
Out of 2052 (9.62, '%') clients with a loan arrear, having 2 children, do not return the loan on time 9.45 %
Out of 330 (1.55, '%') clients with a loan arrear, having 3 children, do not return the loan on time 8.18 %
Out of 41 (0.19, '%') clients with a loan arrear, having 4 children, do not return the loan on time 9.76 %
Out of 9 (0.04, '%') clients with a loan arrear, having 5 children, do not return the loan on time 0.0 %


Childless clients are more likely than clients with one or two children to pay on time. Clients with three or more children make up ~1.78% of all clients and do not significantly affect the conclusion.

### Is there a relationship between marital status and repayment of the loan on time?

In [24]:
for i in data['family_status'].unique():
    family_debt = round(100*data['debt'].loc[data['family_status']== i].mean(),2)
    clients_groped = len(data.loc[data['family_status']==i])
    clients_share = round(100*len(data.loc[data['family_status']==i])/len(data),2),'%' 
    print('Out of',clients_groped,clients_share,'clients with marital status "',i,'"'' share of overdue loans is equal to',family_debt,'%')

Out of 12261 (57.48, '%') clients with marital status " женат / замужем " share of overdue loans is equal to 7.56 %
Out of 4134 (19.38, '%') clients with marital status " гражданский брак " share of overdue loans is equal to 9.31 %
Out of 951 (4.46, '%') clients with marital status " вдовец / вдова " share of overdue loans is equal to 6.62 %
Out of 1189 (5.57, '%') clients with marital status " в разводе " share of overdue loans is equal to 7.06 %
Out of 2796 (13.11, '%') clients with marital status " Не женат / не замужем " share of overdue loans is equal to 9.76 %


Clients with experience of married life are more likely to pay on time

### Is there a relationship between income level and loan repayment on time?

In [25]:
for i in data['total_income_category'].unique():
    client_income_category = round(100*data['debt'].loc[data['total_income_category']== i].mean(),2)
    clients_groped = len(data.loc[data['total_income_category']==i])
    clients_share = round(100*len(data.loc[data['total_income_category']==i])/len(data),2),'%'
    print("Out of",clients_groped,clients_share, "clients in income group" ,i, "share of arrears:" , client_income_category,'%')

Out of 5014 (23.51, '%') clients in income group B share of arrears: 7.06 %
Out of 15921 (74.64, '%') clients in income group C share of arrears: 8.5 %
Out of 349 (1.64, '%') clients in income group D share of arrears: 6.02 %
Out of 22 (0.1, '%') clients in income group E share of arrears: 9.09 %
Out of 25 (0.12, '%') clients in income group A share of arrears: 8.0 %


Clients from group B pay on time more often than clients from group C. Clients from groups: A, E and D make up ~1.86% of all clients and do not significantly affect the conclusion.

###  How do different purposes of a loan affect its repayment on time?

In [26]:
for i in data['purpose_category'].unique():
    debt_category = round(100*data['debt'].loc[data['purpose_category']== i].mean(),2)
    clients_groped = len(data.loc[data['purpose_category']==i])
    clients_share = round(100*len(data.loc[data['purpose_category']==i])/len(data),2),'%'
    print(clients_groped,clients_share,"the client took out a loan for '",i,"' share of arrears:" ,debt_category,'%')
    
    

10751 (50.4, '%') the client took out a loan for ' операции с недвижимостью ' share of arrears: 7.26 %
4279 (20.06, '%') the client took out a loan for ' операции с автомобилем ' share of arrears: 9.35 %
3988 (18.7, '%') the client took out a loan for ' получение образования ' share of arrears: 9.25 %
2313 (10.84, '%') the client took out a loan for ' проведение свадьбы ' share of arrears: 7.91 %


Clients who take out loans for real estate or a wedding often pay on time

## Possible causes of gaps in the source data.

Gaps could appear due to the fact that the bank, for example, requires less data for short-term loans or up to a certain amount, or it is also possible that this is a technical/human error. It would be a good idea to clarify the omissions from the source.

## Why fill in the blanks with the median is the best solution for quantitative variables

Because this work presents employment type in a very general way, the average may skew the figures.

# Stage 3: Study results

* Thus, a client from income group 'B' who is in a married relationship and without children, who has taken out a loan for real estate or a wedding, is potentially reliable

* The majority of the bank's clients (66.06%) are people without children, who have the lowest percentage of late payments (7.54%).

* Also, the group (67.5%) of the most reliable clients includes people with experience in married life, whose average percentage of late payments is 7.08%.

* The client's income is from 200,000 to 1,000,000 (category B) reduces the likelihood of late payment. The non-return rate in this group is 7.06%

* Loans for real estate (50.4%) and weddings (10.84%) are the most reliable. The percentage of late payments is 7.26% and 7.91%, respectively.

* Thus, a client from income group 'B' who is in a married relationship and without children, who has taken out a loan for real estate or a wedding, is potentially reliable

* During the work, gaps in the values of length of service and monthly income were filled in. Blanks are filled with the median value by employment type.